extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentMover.java

```diff
@@ -193,4 +143,110 @@ public Void call() throws Exception
+      String s3Bucket,
+      String targetS3Bucket,
+      String s3Path,
+      String targetS3Path,
+      String copyMsg
+  ) throws ServiceException, IOException, SegmentLoadingException
+  {
+    if (s3Bucket.equals(targetS3Bucket) && s3Path.equals(targetS3Path)) {
+      log.info("No need to move file[s3://%s/%s] onto itself", s3Bucket, s3Path);
+      return;
+    }
+    if (s3Client.isObjectInBucket(s3Bucket, s3Path)) {
+      final S3Object[] list = s3Client.listObjects(s3Bucket, s3Path, "");
+      if (list.length == 0) {
+        // should never happen
+        throw new ISE("Unable to list object [s3://%s/%s]", s3Bucket, s3Path);
+      }
+      final S3Object s3Object = list[0];
+      if (s3Object.getStorageClass() != null &&
+          s3Object.getStorageClass().equals(S3Object.STORAGE_CLASS_GLACIER)) {
+        throw new ServiceException(StringUtils.format(
+            "Cannot move file[s3://%s/%s] of storage class glacier, skipping.",
+            s3Bucket,
+            s3Path
+        ));
+      } else {
+        log.info("Moving file %s", copyMsg);
+        final S3Object target = new S3Object(targetS3Path);
+        if (!config.getDisableAcl()) {
+          target.setAcl(GSAccessControlList.REST_CANNED_BUCKET_OWNER_FULL_CONTROL);
+        }
+        s3Client.copyObject(
+            s3Bucket,
+            s3Path,
+            targetS3Bucket,
+            target,
+            false
+        );
+        if (!s3Client.isObjectInBucket(targetS3Bucket, targetS3Path)) {
+          throw new IOE(
+              "After copy was reported as successful the file doesn't exist in the target location [%s]",
+              copyMsg
+          );
+        }
+        deleteWithRetriesSilent(s3Bucket, s3Path);
+        log.debug("Finished moving file %s", copyMsg);
+      }
+    } else {
+      // ensure object exists in target location
+      if (s3Client.isObjectInBucket(targetS3Bucket, targetS3Path)) {
+        log.info(
+            "Not moving file [s3://%s/%s], already present in target location [s3://%s/%s]",
+            s3Bucket, s3Path,
+            targetS3Bucket, targetS3Path
+        );
+      } else {
+        throw new SegmentLoadingException(
+            "Unable to move file %s, not present in either source or target location",
+            copyMsg
+        );
+      }
+    }
+  }
+
+  private void deleteWithRetriesSilent(final String s3Bucket, final String s3Path)
+  {
+    try {
+      deleteWithRetries(s3Bucket, s3Path);
+    }
+    catch (Exception e) {
+      log.error(e, "Failed to delete file [s3://%s/%s], giving up", s3Bucket, s3Path);
+    }
+  }
+
+  private void deleteWithRetries(final String s3Bucket, final String s3Path) throws Exception
+  {
+    RetryUtils.retry(
+        (Callable<Void>) () -> {
+          try {
+            s3Client.deleteObject(s3Bucket, s3Path);
+            if (s3Client.isObjectInBucket(s3Bucket, s3Path)) {
+              throw new IOException(StringUtils.format(
```

Comment:
is this guaranteed to be the case in S3?

Reply:
No, it isn't, removed